talisker/gunicorn.py

```diff
@@ -82,8 +84,63 @@ class GunicornMetric:
     )
 
 
+# We add a synthetic signal, SIGCUSTOM, to gunicorn's known signals. This
+# allows us to get gunicorn to process the effects of this signal in the
+# arbiter's main loop, rather than within the limited context of the signal
+# handlers.  This makes worker clean up serialized and normal python code.
+
+DEAD_WORKERS = deque()  # storage for recording dead pids
+
+
+def handle_custom():
+    """Handler for a fake 'signal', to be called from the arbiter's main loop.
+
+    This performs the aggregation of prometheus metrics files for dead workers,
+    in a serialized and safe manner.
+    """
+    pid = None
+    while DEAD_WORKERS:
+        try:
+            pid = DEAD_WORKERS.popleft()
+            logger.info('cleaning up prometheus metrics', extra={'pid': pid})
+            talisker.metrics.prometheus_cleanup_worker(pid)
+        except Exception:
+            # we should never fail at cleaning up
+            logger.exception(
+                'failed to cleanup prometheus worker files',
+                extra={'pid': pid},
+            )
+
+
+def gunicorn_on_starting(arbiter):
+    """Gunicorn on_starging server hook.
+
+    Sets up the fake signal and handler on the arbiter instance.
+    """
+    arbiter.SIG_NAMES['SIGCUSTOM'] = 'custom'
```

Comment:
`SIGCUSTOM`, `handle_custom()`, etc sound a bit too generic to me. But I don't have a better alternative so I'll shut up :-)

Reply:
Yeah, I left it generic as it's just a generic way to hook into gunicorn's default main loop signal processing. We might very well use it for something other than prometheus clean up in the future.

SIGTALISKER maybe?

It's doesn't actually need a SIG prefix, and string will do, but I felt it should conform.